momepy/preprocessing.py

```diff
@@ -740,3 +745,368 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+        [
+            pd.DataFrame.from_dict(
+                nodes_to_merge_dict, orient="index", columns=["cluster"]
+            ),
+            nodes_df[[x_col, y_col]],
+        ],
+        axis=1,
+        join="inner",
+    )
+
+    # The two node attributes we need for the clusters are the position of the cluster centroids. Those are obtained by averaging
+    #  the x and y columns. We also add attribtues referring to the original osmids in every cluster:
+    cluster_centroids_df = nodes_to_merge_df.groupby("cluster").mean()
+    cluster_centroids_df["simplified"] = True
+    cluster_centroids_df["original_osmids"] = cluster_centroids_df.index.map(
+        components_dict
+    )
+    cluster_nodes_list = list(cluster_centroids_df.to_dict("index").items())
+
+    # Rebuild edges if necessary:
+    if rebuild_graph:
+        graph.graph["approach"] = "primal"
+        edges_gdf = mm.nx_to_gdf(graph, points=False, lines=True)
+        simplified_edges = _get_rebuilt_edges(
+            edges_gdf,
+            nodes_to_merge_dict,
+            cluster_centroids_df,
+            method=rebuild_edges_method,
+            tolerance=tolerance,
+            edge_from_col=edge_from_col,
+            edge_to_col=edge_to_col,
+        )
+
+    # Replacing the collapsed nodes with centroids and adding edges:
+    simplified_graph = graph.copy()
+    simplified_graph.remove_nodes_from(
+        nodes_to_merge_df.index
+    )  # removing nodes also removes edges!
+    simplified_graph.add_nodes_from(cluster_nodes_list)
+
+    if rebuild_graph:
+        simplified_graph.add_edges_from(simplified_edges)
+
+    return simplified_graph
+
+
+def _get_rebuilt_edges(
+    edges_gdf,
+    nodes_dict,
+    cluster_df,
+    method="linear",
+    tolerance=30,
+    edge_from_col="from",
+    edge_to_col="to",
+):
+    """
+    Rebuilds edge given (potential) new origin and destinations through a linear method: current line is linearly extended
+     to and from the new endpoints.
+
+    :param geometry: shapely.LineString, current edge
+    :param new_origin: shapely.Point or None, new origin point or None if origin remains the same
+    :param new_destination: shapely.Point or None, new destination point or None if destination remains the same
+
+    return list of edges in format (origin_id, destination_id, edge_data)
+    """
+    # Determine what endpoints were made into clusters:
+    edges_gdf["origin_cluster"] = edges_gdf[edge_from_col].apply(
+        lambda u: nodes_dict[u] if u in nodes_dict else -1
+    )
+    edges_gdf["destination_cluster"] = edges_gdf[edge_to_col].apply(
+        lambda v: nodes_dict[v] if v in nodes_dict else -1
+    )
+
+    # Determine what edges need to be simplified (either between diff. clusters or self-loops in a cluster):
+    edges_gdf.query(
+        "origin_cluster != destination_cluster or (('"
+        + edge_to_col
+        + "' == '"
+        + edge_from_col
+        + "') and origin_cluster >= 0)",
```

Comment:
This is probably just a personal preference, but could you make this an f-string instead of this?